File: système_solaire/systeme_solaire.cpp
#include <iostream>
#include <vector>
#include <cmath>

// Structure pour représenter un corps céleste
struct CorpsCeleste {
    std::string nom;
    double masse;
    double positionX, positionY;
    double vitesseX, vitesseY;
};

// Fonction pour calculer la force gravitationnelle
double calculerForceGravitationnelle(double masse1, double masse2, double distance) {
    const double G = 6.67430e-11; // Constante gravitationnelle
    return G * masse1 * masse2 / (distance * distance);
}

// Fonction pour calculer la période orbitale selon la troisième loi de Kepler
double calculerPeriodeOrbitale(double distance, double masseSoleil) {
    const double G = 6.67430e-11;
    return 2 * M_PI * sqrt((distance * distance * distance) / (G * masseSoleil));
}

void caclulerForceGravitationnelle(double masse1, double masse2, double distance) {
    double force = calculerForceGravitationnelle(masse1, masse2, distance);
    std::cout << "Force gravitationnelle: " << force << " N" << std::endl;
}

void calculerVitesseOrbitale(double distance, double masseSoleil) {
    const double G = 6.67430e-11;
    double vitesse = sqrt(G * masseSoleil / distance);
    std::cout << "Vitesse orbitale: " << vitesse << " m/s" << std::endl;
}


int main() {
    // Initialisation des corps célestes (exemple simple avec Soleil et Terre)
    std::vector<CorpsCeleste> systemeSolaire;
    
    CorpsCeleste soleil = {"Soleil", 1.989e30, 0.0, 0.0, 0.0, 0.0};
    CorpsCeleste mercure = {"Mercure", 3.285e23, 5.791e10, 0.0, 0.0, 4.8e4};
    CorpsCeleste venus = {"Venus", 4.867e24, 1.082e11, 0.0, 0.0, 3.5e4};
    CorpsCeleste terre = {"Terre", 5.972e24, 1.496e11, 0.0, 0.0, 2.978e4};
    CorpsCeleste mars = {"Mars", 6.39e23, 2.279e11, 0.0, 0.0, 2.41e4};
    CorpsCeleste jupiter = {"Jupiter", 1.898e27, 7.785e11, 0.0, 0.0, 1.3e4};
    CorpsCeleste saturne = {"Saturne", 5.683e26, 1.433e12, 0.0, 0.0, 9.7e3};
    CorpsCeleste uranus = {"Uranus", 8.681e25, 2.877e12, 0.0, 0.0, 6.8e3};
    CorpsCeleste neptune = {"Neptune", 1.024e26, 4.503e12, 0.0, 0.0, 5.4e3};
    
    systemeSolaire.push_back(soleil);
    systemeSolaire.push_back(mercure);
    systemeSolaire.push_back(venus);
    systemeSolaire.push_back(terre);
    systemeSolaire.push_back(mars);
    systemeSolaire.push_back(jupiter);
    systemeSolaire.push_back(saturne);
    systemeSolaire.push_back(uranus);
    systemeSolaire.push_back(neptune);

    // Boucle de simulation (à compléter)
    double dt = 3600; // Pas de temps en secondes (1 heure)
    int iterations = 1000;
    
    for (int i = 0; i < iterations; ++i) {
        caclulerForceGravitationnelle(soleil.masse, mercure.masse, 5.791e10);
        calculerForceGravitationnelle(soleil.masse, venus.masse, 1.082e11);
        calculerForceGravitationnelle(soleil.masse, terre.masse, 1.496e11);    
        calculerForceGravitationnelle(soleil.masse, mars.masse, 2.279e11);
        calculerForceGravitationnelle(soleil.masse, jupiter.masse, 7.785e11);
        calculerForceGravitationnelle(soleil.masse, saturne.masse, 1.433e12);
        calculerForceGravitationnelle(soleil.masse, uranus.masse, 2.877e12);
        calculerForceGravitationnelle(soleil.masse, neptune.masse, 4.503e12);

        claculerPeriodeOrbitale(5.791e10, soleil.masse);
        calculerPeriodeOrbitale(1.082e11, soleil.masse);
        calculerPeriodeOrbitale(1.496e11, soleil.masse);
        calculerPeriodeOrbitale(2.279e11, soleil.masse);
        calculerPeriodeOrbitale(7.785e11, soleil.masse);
        calculerPeriodeOrbitale(1.433e12, soleil.masse);
        calculerPeriodeOrbitale(2.877e12, soleil.masse);
        calculerPeriodeOrbitale(4.503e12, soleil.masse);

    }
    
    return 0;
}

File: système_solaire/systeme_solaire.py
import math
import numpy as np

noms = ["Soleil", "Mercure", "Venus", "Terre", "Mars", "Jupiter", "Saturne", "Uranus", "Neptune"]

CorpsCeleste = np.array([
    [1.989e30, 0.0, 0.0, 0.0, 0.0],
    [3.285e23, 5.791e10, 0.0, 0.0, 4.8e4],
    [4.867e24, 1.082e11, 0.0, 0.0, 3.5e4],
    [5.972e24, 1.496e11, 0.0, 0.0, 2.978e4],
    [6.39e23, 2.279e11, 0.0, 0.0, 2.41e4],
    [1.898e27, 7.785e11, 0.0, 0.0, 1.3e4],
    [5.683e26, 1.433e12, 0.0, 0.0, 9.7e3],
    [8.681e25, 2.877e12, 0.0, 0.0, 6.8e3],
    [1.024e26, 4.503e12, 0.0, 0.0, 5.4e3]
], dtype=float)

#1. Masse de la planète (en kg) | 2. Distance X du Soleil | 3. Distance Y du Soleil | 4. Vitesse X | 5. Vitesse Y | 6. Diamètre de la planète (en km)

CorpsCelesteTest = np.array([
    [1.989e30, 0.0, 0.0, 0.0, 0.0, 13927e2],
    [3.285e23, 69.8e6, 0.0, 0.0, 4.8e4, 48e2],
    [4.867e24, 108.9e6, 0.0, 0.0, 3.5e4, 121e2],
    [5.972e24, 152.1e6, 0.0, 0.0, 2.978e4, 127e2],
    [6.39e23, 249.2e6, 0.0, 0.0, 2.41e4, 67e2],
    [1.898e27, 816.6e6, 0.0, 0.0, 1.3e4, 1398e2],
    [5.683e26, 1514e6, 0.0, 0.0, 9.7e3, 1164e2],
    [8.681e25, 3003e6, 0.0, 0.0, 6.8e3, 507e2],
    [1.024e26, 4545e6, 0.0, 0.0, 5.4e3, 492e2]
], dtype=float)

dt = 3600
iterations = 0
distance_max = 7.375e9 #distance de l'Aphélie de Pluton

PositionX = CorpsCeleste[:, 3]
PositionY = CorpsCeleste[:, 4]


def ForceGravitationnelle(masse_planet, masse_soleil, dx) :
    force = calculerForceGravitationnelle(masse_planet, masse_soleil, dx)
    return force

def calculerForceGravitationnelle(masse_planet, masse_soleil, dx) :
    G = 6.67430e-11
    return G * masse_planet * masse_soleil / dx**2

planets = [
    ("Mercure", 1),
    ("Venus", 2),
    ("Terre", 3),
    ("Mars", 4),
    ("Jupiter", 5),
    ("Saturne", 6),
    ("Uranus", 7),
    ("Neptune", 8)
]

for i in range(iterations):
    print("Iteration", i + 1)
    print("-----------------------------")
    for name, idx in planets:
        m_p = CorpsCeleste[idx][0]
        m_s = CorpsCeleste[0][0]
        dx = float((CorpsCeleste[idx][1]))
        Fx = ForceGravitationnelle(m_p, m_s, dx)
        print("Soleil - " + name + ": Force gravitationnelle X :", Fx)
        ax = Fx / m_p
        CorpsCeleste[idx][3] = CorpsCeleste[idx][3]
        CorpsCeleste[idx][3] += ax * dt  # update vx
        CorpsCeleste[idx][1] += CorpsCeleste[idx][3] * dt  # update x